### PR TITLE
Fix an issue accessing invalid pointer

### PR DESCRIPTION
**Pull request by cyx1231st**
posix_socket_impl lives in stack, it will be destroyed right after the `connect()` and `find_port_and_connect()` returned.

### CODE REVIEW

**Comment by cyx1231st** [comment]
@tchaikov @rzarzynski

File: src/net/posix-stack.cc [hunk 1]
```diff
@@ -159,10 +159,10 @@ class posix_socket_impl final : public socket_impl {
     future<> find_port_and_connect(socket_address sa, socket_address local, transport proto = transport::TCP) {
         static thread_local std::default_random_engine random_engine{std::random_device{}()};
         static thread_local std::uniform_int_distribution<uint16_t> u(49152/smp::count + 1, 65535/smp::count - 1);
-        return repeat([this, sa, local, proto, attempts = 0, requested_port = ntoh(local.as_posix_sockaddr_in().sin_port)] () mutable {
+        return repeat([fd = _fd, sa, local, proto, attempts = 0, requested_port = ntoh(local.as_posix_sockaddr_in().sin_port)] () mutable {
```
Comment (rzarzynski) [inline src/net/posix-stack.cc:162]: OK, `_fd` is locally ref counted.

**Review by rzarzynski — APPROVED** [review]

**Review by tchaikov — COMMENTED** [review]
`posix_socket_impl` is wrapped by `std::unique_ptr<net::socket_impl>` returned by `network_stack::socket()` as a member variable of `socket`.

so i think `posix_socket_impl` is not allocated on stack.

but i agree with you that we cannot capture `this` for capturing `_fd`, because, in the expression of `socket().connect(sa, local, proto)`, `socket()` returns a temporary object whose will be destructed when`connect(sa, local, proto)` returns.

from C++'s perspective, `future<connected_socket>` is returned right away, and the compiler believes that we have done with the return value of `socket()` .

but from seastar's perspective, the whatever action will be repeated by `find_port_and_connect()` has been wrapped up in a task which is in turn scheduled to be run by reactor later. and the chained  continuations will be evaluated after the future returned by `find_port_and_connect()` is available.

that's why we cannot assume that the temporary object returned by `socket()` can be still alive when the continuations scheduled by `socket().connect(sa, local, proto)` is evaluated.


so, could you revise the commit message. and send this patch to upstream, see 

- https://github.com/scylladb/seastar/blob/master/doc/contributing.md and 
- https://github.com/scylladb/scylla/wiki/Formatting-and-sending-patches

for more details on contributing to seastar.

**Comment by cyx1231st** [comment]
There is an on-going fix in seastar-dev:
https://groups.google.com/d/msg/seastar-dev/HJUqaNEX4G0/WPDDtKLgDQAJ

**Comment by tchaikov** [comment]
i think your fix is better. `do_with()` creates an extra continuation for holding the state.

**Comment by cyx1231st** [comment]
please refer to https://groups.google.com/d/msg/seastar-dev/P0ytNGDq6DI/znW_53kUDwAJ

**Comment by cyx1231st** [comment]
fixed by https://github.com/ceph/seastar/commit/5e399906d1efa9e5ca9d121d09c6d2b1cdc41404

**Pull request closed without merging**